modepy/modes.py

```diff
@@ -657,33 +658,82 @@ def grad_simplex_best_available_basis(dims, n):
 class _TensorProductBasisFunction:
     # multi_index is here just for debugging.
 
-    def __init__(self, multi_index, per_dim_functions):
+    def __init__(self, multi_index, functions, *, dims_per_function):
+        assert len(dims_per_function) == len(functions)
+
         self.multi_index = multi_index
-        self.per_dim_functions = per_dim_functions
+        self.functions = functions
+
+        self.dims_per_function = dims_per_function
+        self.ndims = sum(self.dims_per_function)
 
     def __call__(self, x):
         result = 1
-        for iaxis, per_dim_function in enumerate(self.per_dim_functions):
-            result = result * per_dim_function(x[iaxis])
+        d = 0
+
+        for n, function in zip(self.dims_per_function, self.functions):
+            result *= function(x[d:d + n])
+            d += n
 
         return result
 
+    def __repr__(self):
+        return (f"{type(self).__name__}(mi={self.multi_index}, "
+                f"dims={self.dims_per_function}, functions={self.functions})")
+
 
 class _TensorProductGradientBasisFunction:
     # multi_index is here just for debugging.
 
-    def __init__(self, multi_index, per_dim_derivatives):
+    def __init__(self, multi_index, derivatives, *, dims_per_function):
+        assert len(dims_per_function) == len(derivatives[0])
+
         self.multi_index = multi_index
-        self.per_dim_derivatives = tuple(per_dim_derivatives)
+        self.derivatives = tuple(derivatives)
+
+        self.dims_per_function = dims_per_function
+        self.ndims = sum(self.dims_per_function)
 
     def __call__(self, x):
-        result = [1] * len(self.per_dim_derivatives)
-        for ider, per_dim_functions in enumerate(self.per_dim_derivatives):
-            for iaxis, per_dim_function in enumerate(per_dim_functions):
-                result[ider] *= per_dim_function(x[iaxis])
+        # NOTE: we're trying to do a tensor product gradient for something like
+        #
+        #       P_1(x0, x1) * P_2(x2) * P_3(x3, x4, x5) * ...
+        #
+        # (here `dims_per_function` would be `(2, 1, 3)`), where each
+        # `self.derivatives[i]` contains the derivatives of `P_i` and the
+        # remaining functions as is (in the same order as above). Then, to
+        # get the derivatives with respect to, e.g., `(x3, x4, x5)`, we
+        # loop over the `components` of the gradient of `P_3` as
+        #
+        #   result[d + 0] = P_1 * P_2 * dP_3/dx_3 * ...
+        #   result[d + 1] = P_1 * P_2 * dP_3/dx_4 * ...
+        #   result[d + 2] = P_1 * P_2 * dP_3/dx_5 * ...
```

Comment:
I tried to describe it a bit better in 
d823e67 but it should be more or less the same as what was in there before (see [here](https://github.com/inducer/modepy/blob/321b2747f86b0a2d111dd80b09a04408ede1c567/modepy/modes.py#L681-L683)).

Basically before, each `per_dim_derivatives` contained a list of
```
(f_1, ..., df_i, ..., f_n)
```
and there was not much to do besides evaluate each of them because they were all scalar. Now, `df_i` can return a tuple, with one entry per variable it depends on, so there's a bit more accounting for that.

Reply:
That helps, thanks!